subprojects/core/src/integTest/groovy/org/gradle/api/tasks/NestedInputIntegrationTest.groovy

```diff
@@ -658,6 +654,86 @@ class NestedInputIntegrationTest extends AbstractIntegrationSpec {
         failure.assertHasCause("Cycles between nested beans are not allowed. Cycle detected between: 'nested' and 'nested.nested'.")
     }
 
+    def "duplicate names in nested iterable cause the build to fail"() {
+        buildFile << taskWithNestedInput()
+        buildFile << namedBeanClass()
+        buildFile << """
+            myTask.nested = [new NamedBean('name', 'value1'), new NamedBean('name', 'value2')]           
+        """
+
+        expect:
+        fails "myTask"
+        failure.assertHasDescription("Could not determine the dependencies of task ':myTask'.")
+        failure.assertHasCause("Nested iterables can only contain beans with unique names. Duplicate name: 'nested.name'.")
+    }
+
+    def "task with named nested beans is incremental"() {
+        buildFile << taskWithNestedInput()
+        buildFile << namedBeanClass()
+        buildFile << """                                   
+            myTask.nested = [new NamedBean(project.property('namedName'), 'value1'), new NamedBean('name', 'value2')]           
+        """
+        def taskPath = ':myTask'
+
+        when:
+        run taskPath, '-PnamedName=name1'
+        then:
+        executedAndNotSkipped taskPath
+
+        when:
+        run taskPath, '-PnamedName=name1'
+        then:
+        skipped taskPath
+
+        when:
+        run taskPath, '-PnamedName=different', '--info'
+        then:
+        executedAndNotSkipped taskPath
+        output.contains("Input property 'nested.different.class' has been added for task ':myTask'")
+        output.contains("Input property 'nested.name1.class' has been removed for task ':myTask'")
```

Comment:
Would be nice if we could report the property name without `.class` here, but maybe do that in a separate issue...

Reply:
Why do we add the `.class` extension in the first place? Couldn't we just add `nested.name1` as a property which has the class as value?
Then we don't have to sanitize the property name [here](https://github.com/gradle/gradle/blob/964b0e167746880318d8d41082bed383d0b244e2/subprojects/core/src/main/java/org/gradle/caching/internal/tasks/DefaultTaskOutputCachingBuildCacheKeyBuilder.java#L107), either.

However we fix this, I guess we should do it in a separate PR.

Reply:
👍